### PR TITLE
WFLY-7085 - Server should verify EJB business methods during deployment and log a warning

### PR DESCRIPTION
**Pull request by rpelisse**
Issue: https://issues.jboss.org/browse/JBEAP-5947
Upstream Issue: https://issues.jboss.org/browse/WFLY-11380 (but is a followup work from https://issues.jboss.org/browse/WFLY-10048)

### CODE REVIEW

**Comment by rpelisse** [comment]
For the record, the failure on Windows seems unrelated to the changes:

**Comment by rpelisse** [comment]
retest this please

File: ejb3/src/main/java/org/jboss/as/ejb3/component/EJBValidationConfigurator.java [hunk 1]
```diff
@@ -52,27 +50,18 @@ private EJBValidationConfigurator() {
...
-        }
-        if(!Modifier.isPublic(configuration.getComponentClass().getModifiers())) {
-            throw EjbLogger.ROOT_LOGGER.ejbMustBePublicClass(description.getComponentName(), description.getComponentClassName());
-        }
-        if(Modifier.isFinal(configuration.getComponentClass().getModifiers())) {
-            throw EjbLogger.ROOT_LOGGER.ejbMustNotBeFinalClass(description.getComponentName(), description.getComponentClassName());
-        }
+
+        EjbValidationsUtil.verifyEjbClassAndDefaultConstructor(ctor, configuration.getComponentClass().getEnclosingClass(), noInterface, description.getComponentName(), description.getComponentClassName(), configuration.getComponentClass().getModifiers());
+        EjbValidationsUtil.verifyEjbMethodAreNotFinalNorStatic(configuration.getComponentClass().getDeclaredMethods(),description.getComponentClassName());
```
Comment (kabir) [inline ejb3/src/main/java/org/jboss/as/ejb3/component/EJBValidationConfigurator.java]: @rpelisse @mjurc The Jira states: "The ejb 3.2 core specification state in chapter 4.9.6 that business methods should not be declared as final or static.". Since we're doing class.getDeclaredMethods() here we are getting all methods of the class itself, and verifying those. To me that looks wrong as getDeclaredMethods() will return all private methods as well, which are not business methods. Also, does a method being public necessarily need to be a business method (e.g. if a local/remote interfaces are defined and the method is not declared there)? Finally, the ejb class could inherit some business methods from a super class, so we should probably check those as well?

File: ejb3/src/main/java/org/jboss/as/ejb3/component/EJBValidationConfigurator.java [hunk 2]
```diff
@@ -52,27 +50,18 @@ private EJBValidationConfigurator() {
...
+        for ( Class<?> interfaceClass : configuration.getComponentClass().getInterfaces())
+            EjbValidationsUtil.verifyEjbMethodAreNotFinalNorStatic(interfaceClass.getDeclaredMethods(), interfaceClass.getCanonicalName());
```
Comment (kabir) [inline ejb3/src/main/java/org/jboss/as/ejb3/component/EJBValidationConfigurator.java]: I wasn't sure, but a quick test shows me that neither static nor final are allowed on interfaces

File: ejb3/src/main/java/org/jboss/as/ejb3/deployment/processors/MessageDrivenComponentDescriptionFactory.java [hunk 3]
```diff
@@ -103,7 +103,7 @@ private void processMessageBeans(final DeploymentUnit deploymentUnit, final Coll
         for (final AnnotationInstance messageBeanAnnotation : messageBeanAnnotations) {
             final AnnotationTarget target = messageBeanAnnotation.target();
             final ClassInfo beanClassInfo = (ClassInfo) target;
-            if (! MdbValidationsUtil.assertMDBClassValidity(beanClassInfo).isEmpty() ) {
+            if (! EjbValidationsUtil.assertEjbClassValidity(beanClassInfo).isEmpty() ) {
```
Comment (kabir) [inline ejb3/src/main/java/org/jboss/as/ejb3/deployment/processors/MessageDrivenComponentDescriptionFactory.java]: Since this seems to only be used for MDBs, why the rename?

**Review by kabir — CHANGES_REQUESTED** [review]
@rpelisse I am not familiar with the EJB implementation, but perhaps the validation should be done once the business methods have been determined instead?

File: ejb3/src/main/java/org/jboss/as/ejb3/component/EJBValidationConfigurator.java (hunk 2 above)
Comment (rpelisse) [reply ejb3/src/main/java/org/jboss/as/ejb3/component/EJBValidationConfigurator.java]: @kabir excellent point, I've removed that.

File: ejb3/src/main/java/org/jboss/as/ejb3/component/EJBValidationConfigurator.java (hunk 1 above)
Comment (rpelisse) [reply ejb3/src/main/java/org/jboss/as/ejb3/component/EJBValidationConfigurator.java]: >  To me that looks wrong as getDeclaredMethods() will return all private methods as well, which are not business methods. Also, does a method being public necessarily need to be a business method (e.g. if a local/remote interfaces are defined and the method is not declared there)?

I've added some logic to adress that.

> Finally, the ejb class could inherit some business methods from a super class, so we should probably check those as well?

I would not got that far. I think validation should take the low hanging fruit, but starting to explore inheritance and such things sounds a bit like overkill. Also, I would rather introduce the current set of changes, see if it breaks and leads to issues and then maybe go deeper and try to validate more advanced use case. (IMHO)

File: ejb3/src/main/java/org/jboss/as/ejb3/deployment/processors/MessageDrivenComponentDescriptionFactory.java (hunk 3 above)
Comment (rpelisse) [reply ejb3/src/main/java/org/jboss/as/ejb3/deployment/processors/MessageDrivenComponentDescriptionFactory.java]: @kabir Because this PR aims at extending the validation did to Mdb to Ejb! :)

**Comment by rpelisse** [comment]
@kabir friendly reminder ;) - I'm up to do any fix you see fit, I just want to be sure the PR is not in a stale state!

**Comment by jamezp** [comment]
I think we need a new JIRA for this as it's rather confusing. The commit message indicates WFLY-7085 yet the PR descriptions links to JBEAP-5947 and WFLY-10048. As someone who's not familiar with the EJB implementation it would be nice to know what exactly the issue is so the fix can be validated.

**Comment by rpelisse** [comment]
@jamezp I can create a new JIRA issue if you prefer but I think it might create more confusion. The issue at hand in this PR is actually pretty simple and the commit message is spot on. It's more the history of issues that is a bit confusing. 

Just to recap: the EJB specification has some limitations on what can be done inside an MDB class - ex: onMessage() method name has a specific meaning with MDB so you can used this name freely. There was some validation in place in Wildfly, to check that MDB were compliant, but the coverage was partial. 

During some messaging testing this partial coverage was pointed out and the validation code was reworked and enhanced. So now, if a MDB is not complaint, appropriate warning messages are being logged. This is the all story behind WFLY-10048. 

Following this, it was pointed out that the specification has also constraints on other EJB (not only MDB). That is (in short) the content of JBEAP-5947. This PR rework and reuse the validation code done for MDB to extend it to any EJBs. 

That's just it. Honestly, it does not require a deep knowledge of EJB - otherwise I would have certainly not been able to do it :) . To dumb it down, it is a series of "if" checking that the EJB does not have violations against the EJB specification recommendation.

I hope this clarify things a bit. If you prefer, I can create an upstream issue for JBEAP-5947 or reopen the one that was resolved (WFLY-7085) when the MDB part was verified). Let me know! Whatever you feel is better.

**Comment by jamezp** [comment]
I would say at a minimum I think WFLY-7085 should be reopened. However I would still vote for a new JIRA. The reason being that WFLY-7085 was fixed in WildFly 14 and this won't be fixed until WildFly 15 so it would be confusing for people trying to figure out which release this was really fixed in. Having 2 issues linked and a different issue in the commit message is what confuses me most :)

**Comment by rpelisse** [comment]
@jamezp I've created a new JIRA and updated the commit message accordingly. Sorry for the delay, this task somehow "disappeared" from my TODO list! Let me know if you want me to change something else!
